_resources/port1.0/group/mpi-1.0.tcl

```diff
@@ -341,7 +341,7 @@ proc mpi.setup {args} {
         # Clang 15 and 16 only available on 10.7 and later
         if {${os.major} < 11} {
             lappend ::mpi.disabled_compilers \
-                -clang15 -clang16
+                -clang15 -clang16 -clang17
```

Comment:
I have no stakes here, I do not use any Clangs on old systems, but AFAIK we do have Clang 15+ on older Intel after @catap fixed them.

Reply:
all clangs newer than clang-11 are broken on older systems unless the user manually installs a newer linker, as I recall.

Nobody has sorted out how to get a newer linker bootstrapped automatically, ie in the buildbot.

Until someone does sort that out, the newest clang that works with the current default linker that can be installed in clang-11.

Reply:
@kencu as far as I recall @mascguy had tested it on his machines and confirmed that it works well

Reply:
Ken, here the @mascguy confirms that he used some of clang after 11 to build never clang's https://github.com/macports/macports-ports/pull/19791#issuecomment-1719397709

Can you point that doesn't work on 10.5 or 10.6?

Reply:
The TWO times someone tried this, it had to be reverted both times due to failures for users and the buildbots:

https://github.com/macports/macports-ports/commit/3016286b6087f47cf660773c877ab9b70f5896fb

https://github.com/macports/macports-ports/commit/933ee8c94faeec93e9745090fceb310a44f2a8cd

Reply:
I have no idea what @mascguy did or did not see, or how his system was configured, or how your system was configured. ... just that any clang newer than 11 doesn't (yet) work properly before 10.7 without a manual linker update.

Reply:
and of course clang-17 is broken on most older systems because it needs a newer libc++.dylib than most systems have (and newer than the macports-libcxx port offers as well).

that might be fixable one day by adding missing symbols into legacy support, perhaps.

Reply:
Ken, I had encountered at least a few strong indicators that some (all?) build bot isn't clean system. For example it had (has?) installed additional JVM. Also, by some unknown for me reason `cmake` on 10.6 x86_64 returns `CMAKE_SIZEOF_VOID_P` with value different than `8` which I can't reproduce on clean system.

I assume that *something* has been made with that system and if something is failed on it which can't be reproduce on clean system it strong symptom that issue may came from build bot.

Reply:
> I have no idea what @mascguy did or did not see, or how his system was configured, or how your system was configured. ... just that any clang newer than 11 doesn't (yet) work properly before 10.7 without a manual linker update.

The PR Kirill mentioned prior, has all of the detail. But the only thing changed from a stock install, is switching to a newer version of `ld64`. (Which I had already done a few months earlier anyway.)

Reply:
@mascguy : upgrading the linker manually to a newer linker is precisely what is required to make newer clangs work.

@catap : the buildbots are by definition a clean system.

So the toolchains are pretty shaky. Lots of things don’t work right, or aren’t built right at present. Too many hacks. I would suggest folks stop throwing new wrenches into the machine, and work on fixing the many things that already need fixing.

gcc10-bootstrap is not supposed to be built in different arches. It’s supposed to be able to build different arches. Finding out a year in that clang-11-bootstrap can’t build anything universal means nobody ever tried that. The gcc ports that are supposed to universal arm/Intel actually aren’t, so can’t build anything universal. cctools and ld64 need to be updated, and built against newer llvms, and both need a workable bootstrapping plan. 

Things are in a rather sorry state. Things need far more testing before pushing.

Reply:
@kencu 
> gcc10-bootstrap is not supposed to be built in different arches. It’s supposed to be able to build different arches.

What do you mean? GCC should be able to build in different archs (at least same family ones, I am not sure about cross). While in Macports other gcc* ports are indeed broken and do not build as proper universal, `gcc10-bootstrap`, AFAIU, build correctly as multi-arch.

This still leaves open the problem of _using it_ to build other ports as universal (including other gcc*), of course.

> The gcc ports that are supposed to universal arm/Intel actually aren’t

This was broken forever in Macports.

Reply:
I rest my case.

Reply:
> @kencu
> 
> > gcc10-bootstrap is not supposed to be built in different arches. It’s supposed to be able to build different arches.
> 
> What do you mean? 

I am not sure you will follow this, but here it is, again - 

To build Intel code on a powerpc machine, you don't run an intel arch compiler; you run a compiler that generates intel-arch code. To generate i386 code on an Intel system running x86_64, you don't run a compiler that is i368 architecture, you run an x86_64 architecture compiler that generates i386 code..

The current gcc10-bootstrap port is quite broken, although it can build some cross things by accident, with rosetta. Try building arm64 code on an Intel arch machine, for example, to see it fail.

> 
> > The gcc ports that are supposed to universal arm/Intel actually aren’t
> 
> This was broken forever in Macports.

And should be the primary thing for people to fix, as universal building is arguably macports' main advantage over the competition, and yet is hopelessly borked in many ways.


> 
> > the buildbots are by definition a clean system.
> 
> They _should_ be, but whether they _are_ or not is not decided by an imperative.

Given that they are, I think the onus is on you to show they aren't. The example @catap showed above, where the size of a void type was off, is due to the kernel arch being i386 for both 32bit and 64 bit snow leopard. This is a well-known thing we have worked around for 15 years. It has nothing to do with the build bots being supposedly "haunted" and not building cleany.

So the buildbots _are_ clean, but whether you feel they are not can't be decided by an accusation.

Reply:
@kencu Ok, so you expect from a universal compiler to be a native plus cross-compiler (host=a, target=b), while I expect it to be FAT (host=target), is this correct?

These are just different things, neither of which, provided they work, is broken.

What I want is, for example, being able to use the same FAT binary on a G5 when Macports is configured as ppc, or ppc64, or ppc+ppc64 universal. I thought this is also what GCC itself actually does outside of Macports (sorry, cannot check that right now, but will soon, I have Iain branch built as universal on Leopard, I think).
The same FAT compiler should work on 10.6.x (both released and dev builds) for Intel and ppc (by virtue of Rosetta), where ppc slice gonna generate ppc code and Intel slice – x86 code.
It will not work for cross-compiling, of course.

There are cross-compiling tools, and it seems to me that they should not be confused with universal builds per se.

Reply:
@kencu Re buildbots: I nowhere make a claim they are *not* clean systems, my statement was purely logical: something that is supposed to be the case does not prove it is actually the case. I am not particularly interested in investigating buildbots set-ups, it was a side-point.

Reply:
All users expect a compiler labelled universal to build software in the universal arches it claims to support.

Reply:
@kencu This is not the point. When we say a port is universal we mean it can run on either of the archs it is built for. You seem to jump to a rather different definition of universal for a case of a compiler, understanding it as a cross-compiler.

Make it obvious: there are two archs, a and b.
Universal gcc, AFAIU, means it is built for a and b, can run natively on a and b and generate code for a and b.
What you seem to want is a different thing: it can only run natively on a, but generate code for both. In my view, this is a cross-compiler, not a compiler built as universal.

Reply:
In all of MacPorts, when a compiler is marked as universal, we mean first and foremost it can generate code for its universal arches. That is 100% exactly the point.

Having a FAT compiler is possibly desirable, I guess, to avoid confusion, but it is very very bloaty. The clang11bootstrap port comes in at nearly 800 MB. That is pretty stupid, to be honest, and is yet another toolchain problem that will probably need to be addressed.

Having a FAT compiler that can only generate code for the non-native arch by running the non-native arch compiler under rosetta is really quite ridiculous. It is slow, and only works 1/2 the time.

So back to my point -- there are so many toolchain problems on macports right now, that taking a step back and starting a project to actually fix them might be helpful, rather than troweling more cement into the cracks.

Reply:
@kencu There is arguably no point in labeling a non-universal cross-compiler a universal one. It may be the case that more users would rather have a non-FAT cross than a FAT compiler, or vice verse, but it is irrelevant: avoiding misleading definitions is the right thing to do.

> Having a FAT compiler that can only generate code for the non-native arch by running the non-native arch compiler under rosetta is really quite ridiculous.

When you build any other port with `+universal` you expect it to _run_ on every arch natively which is specified in universal archs settings. By the same token, a universal gcc should be a FAT compiler, where every slice generates respective code. It is no more ridiculous than building anything as universal. A user should be able to build `gcc +universal` and use the same binary on any machine. What you suggest will be broken: once you install a x86_64 cross-gcc on an i386 machine, it is just useless, binaries will not run.

Sure enough, we do not want a FAT compiler which generates code for a single arch (which will be a cross-compiler again, but other way round). What is needed is a genuine universal, where each slice generates respective code. What Macports has now, AFAIU, is a fake universal – it pretends to be such, but binaries can only run on a single arch.
gcc10-bootstrap is, ironically, the only gcc which is actually universal in the proper sense. However, the set-up itself remains broken, since there is no way to get to a universal gcc13.

Reply:
So round and round we go, me having actually run the compilers on macports for nearly a decade, and you for -- never.

Having a FAT compiler that can only generate code for the non-native arch by running the non-native arch compiler under rosetta is really quite ridiculous.

Reply:
There is no real need to spend effort trying to prove something as trivial as the fact that you are in Macports much longer – since it was not disputed to begin with, at least not by me.

If only you could find some time during that decade to fix universal builds…